services/common/src/components/forms/FormWrapper.tsx

```diff
@@ -0,0 +1,79 @@
+import React, { FC, useEffect } from "react";
+import { Form } from "antd";
+import { compose } from "@reduxjs/toolkit";
+import { connect, useDispatch, useSelector } from "react-redux";
+import { reduxForm, submit, getFormSubmitErrors, InjectedFormProps, ConfigProps } from "redux-form";
+
+export interface IFormContext {
+  isEditMode: boolean;
+  isModal: boolean;
+}
+
+export const { Provider: FormProvider, Consumer: FormConsumer } = React.createContext<IFormContext>(
+  {
+    isEditMode: true,
+    isModal: false,
+  }
+);
+
+/**
+ * USAGE NOTES:
+ * - pass in params like "touchOnBlur" and "onSubmitSuccess" for reduxForm through reduxFormConfig for custom behaviour
+ */
+interface FormWrapperProps {
+  name: string;
+  reduxFormConfig?: Partial<ConfigProps>;
+  onSubmit: (values) => void | Promise<void>;
+  children: any;
+  isModal?: boolean;
+  loading?: boolean;
+  isEditMode?: boolean;
+  scrollOnToggleEdit?: boolean;
+}
+
+const FormWrapper: FC<FormWrapperProps & InjectedFormProps<any>> = ({
+  isEditMode = true,
+  isModal = false,
+  scrollOnToggleEdit = true,
+  children,
+  ...props
+}) => {
+  const providerValues = {
+    isEditMode,
+    isModal,
+  };
+  const dispatch = useDispatch();
+  const formErrors = useSelector(getFormSubmitErrors(props.name));
+
+  useEffect(() => {
+    if (scrollOnToggleEdit) {
+      window.scrollTo(0, 0);
+    }
+  }, [isEditMode]);
+
+  const handleSubmit = (values) => {
+    dispatch(submit(props.name));
+    if (!formErrors) {
+      props.onSubmit(values);
+    }
+  };
+
+  const formClassName =
+    `common-form common-form-${props.name} form-` + isEditMode ? "edit" : "view";
+
+  return (
+    <FormProvider value={providerValues}>
+      <Form layout="vertical" onFinish={handleSubmit} name={props.name} className={formClassName}>
+        {children}
+      </Form>
+    </FormProvider>
+  );
+};
+
+const mapStateToProps = (_state, ownProps) => ({
+  form: ownProps.name,
+  ...ownProps.reduxFormConfig,
+});
+export default compose(connect(mapStateToProps), reduxForm({}))(FormWrapper as any) as FC<
+  FormWrapperProps
+>;
```

Comment:
I'm wondering, since this wrapper component is a functional one, would it have been possible to use a useSelector hook instead of mapping state to props here?

Reply:
There's some tricky redux logic here (I struggled with it!) that I put in to allow reduxForm to receive a dynamic form name and custom config (ie: you've probably seen touchOnBlur, onSubmitSuccess, etc. All that can be passed in as reduxFormConfig and it'll go into the reduxForm constructor)
ReduxForm actually creates a bunch of components (this screenshot is from a single instance):
![image](https://github.com/bcgov/mds/assets/102187683/4188eb38-ea32-44c6-b6a3-f9d6083c44c5)
The mapStateToProps is what I gotta do to pass those values "up" the tree, and so that reduxForm receives it like props.

Reply:
Ah! I see why you had to do it this way. Cool cool.